### PR TITLE
Natively animated `ShapeSource` for `Point` and `LineString`

### PR DESCRIPTION
**Pull request by naftalibeder**
## Description

This adds two new components, `AnimatedPointSource` and `AnimatedLineSource`.

Currently, animating a point or a line on the map requires sending messages over the bridge many times per second. This is inefficient even when sending small updates (hence why Reanimated exists), but animating especially a long line involves transferring a very large array repeatedly, where potentially most of the coordinates are identical from frame to frame.

These components minimize the messages (both size and frequency) that need to be sent from the React Native app. With `AnimatedPointSource`, you can specify a duration and then simply send coordinate updates, and the source will animate to them. With `AnimatedLineSource`, you can send the entire line string once, and then just update the start ~and end~ (not implemented yet) distances, and the source will animate the line ends accordingly.

Note: I considered augmenting `ShapeSource` instead, but I decided that it would potentially overwhelm that component and make it hard to maintain.

## Checklist

- [x] I have tested this on a device/simulator for each compatible OS
- [x] I updated the documentation with running `yarn generate` in the root folder
- [x] I added/updated a sample - if a new feature was implemented (`/example`)

## Videos

https://github.com/rnmapbox/maps/assets/2423291/3dfb20b5-51f9-4020-b600-ff55ad74fe21

https://github.com/rnmapbox/maps/assets/2423291/5352a0e1-3780-4280-b297-aeb7e3838475

https://github.com/rnmapbox/maps/assets/2423291/89f5cdfc-5387-4b01-862c-29fa22aae9f5

https://github.com/rnmapbox/maps/assets/2423291/a44d0b19-7763-45a1-980f-83ce6e8843ca

### CODE REVIEW

**Review by mfazekas — COMMENTED** [review]
Thanks that is awesome! 

Is there any way we could avoid adding new sources, for this purpose?
I was thinking of something like this:

```swift
class RCTMGLShapeSource : RCTMGLSource {
   ...

   lazy var shapeAnimator : RNMBXShapeAnimator = {
     let shapeAnimator = RNMBXShapeAnimator(onUpdate: {
        self.doUpdate { (style) in
          logged("RCTMGLShapeSource.setShape") {
            try style.updateGeoJSONSource(withId: id, geoJSON: obj)
          }
        }
     })
     return shapeAnimator;
   }()
   
   @objc var animatedShape : String? {
      didSet {
         shapeAnimator.setValue(animatedShape)
      } 
   }
```

But it's just a rough idea, just showing that we don't have to bake in animation logic into ShapeSource, just reference. I'd prefer adding a few lines to ShapeSource, vs creating new classes for this purpose. 

What do you think?

**Comment by naftalibeder** [comment]
Sorry it took so long to get to this.

I generally agree about the value of a single component. I had a few concerns which led me to make separate components:

- Possible disagreement about what it means to animate a line - the particular use case here is animating the beginning and end, but keeping the line string constant. But I could also imagine somebody wanting to actually morph one line into another, which this PR doesn't support, and which would call for an incompatible implementation.
- Having one component support animations for some data types but not others. For example, what would the implementation be for animating a clusterable `FeatureCollection`?
- If we were to abstract `ShapeSource` to handle animation, I think the component and TS glue would need a refactor, particularly in the interface, to prevent it from getting really confusing and enabling nonsensical property combinations. I could see something like

```ts
type ShapeSourceBaseProps = {
  id: string;
  children?: React.ReactElement | React.ReactElement[];
};

type LineSourceProps = {
  readonly type: "LineString";
  shape: GeoJSON.LineString;
  startOffset?: number;
  endOffset?: number;
  animationDuration?: number;
};

type PointSourceProps = {
  readonly type: "Point";
  shape: GeoJSON.Point;
  snapIfDistanceIsGreaterThan?: number;
  animationDuration?: number;
};

type ShapeSourceProps = ShapeSourceBaseProps &
  (LineSourceProps | PointSourceProps);
```

That way, these would be valid:

```jsx
<ShapeSource
  type={'LineString'}
  id={'my-line-source'}
  shape={...}
  startOffset={250}
  endOffset={0}
  animationDuration={1000}
>
  ...
</ShapeSource>

<ShapeSource
  type={'Point'}
  id={'my-point-source'}
  shape={...}
  snapIfDistanceIsGreaterThan={300}
  animationDuration={1000}
>
  ...
</ShapeSource>
```

but these would correctly throw errors:

```jsx
<ShapeSource
  type={'Point'}
  id={'my-point-source'}
  shape={...}
  startOffset={250} // property `startOffset` does not exist
  animationDuration={1000}
>
  ...
</ShapeSource>

<ShapeSource
  type={'PointFeatureCollection'}
  id={'my-points-feature-collection-source'}
  shape={...}
  animationDuration={1000} // property `animationDuration` does not exist
>
  ...
</ShapeSource>
```

Another possibility is going with a generic:

```ts
type LineSourceProps = {
  shape: GeoJSON.LineString;
  startOffset?: number;
  endOffset?: number;
  animationDuration?: number;
};

type PointSourceProps = {
  shape: GeoJSON.Point;
  snapIfDistanceIsGreaterThan?: number;
  animationDuration?: number;
};

type ShapeSourceProps<T = LineSourceProps | PointSourceProps> = {
  id: string;
  shapeProps: T;
  children?: React.ReactElement | React.ReactElement[];
};
```

to be used like:

```jsx
<ShapeSource<LineSourceProps>
  id={'my-line-source'}
  shapeProps={
    shape: ...,
    startOffset: 250,
    endOffset: 0,
  }
  animationDuration={1000}
>
  ...
</ShapeSource>
```

but this wouldn't allow the same runtime type inspection that the first method would give us.

Thoughts?

**Comment by mfazekas** [comment]
What about introducing a parent component, if ShapeSource is placed inside of on AnimatedShape containers then shape property would not be allowed.

```
<AnimatedLineShape
  animationDuration={...}
  ...
>
   <ShapeSource id={...} ...>
</AnimatedLineShape>
```

**Comment by naftalibeder** [comment]
You can probably tell I don't have a lot of time to work on this, so I'm realizing it would be smart to try to find a relatively simple solution.

If you prefer to not have new native components, what about just adding extra props to `ShapeSource`, like `animationDuration` and `snapIfDistanceIsGreaterThan`? Then e.g. `RCTMGLShapeSource` can conditionally add and remove specialized subcomponents, but the native <--> JS interface would still be only `ShapeSource`.

What do you think? I'll try to respond a little more quickly this time :)

**Comment by naftalibeder** [comment]
@mfazekas This is now working on both platforms with no new native files - everything is inside of `ShapeSource`. You can see up-to-date examples in the description above.

Thoughts?

**Comment by mfazekas** [comment]
@naftalibeder thanks much for working on this. 

I've made a POC implementation on the structure described earlier:

```jsx
<AnimatedLineShape
  animationDuration={...}
  shape={...}
  startOffset={...}
  endOffset={...}
  ...
>
   <ShapeSource id={...} ...>
</AnimatedLineShape>
```
See:
https://github.com/rnmapbox/maps/pull/3012

I think it offers the following advantages: 
- we can have multiple animator classes as shapes can be animated multiple ways
- animation logic and shape source logic is separated (this simplifies both usage/docs and implementation)
- does support future structure when a path animator have one  shape source for the enterie line, while the shape source for the end of line

The POC needs to implemented on Android and the real animation functionality implemented in your PR moved, what do you think?

**Comment by naftalibeder** [comment]
@mfazekas Interesting idea. It does have the benefit of keeping the animation component specific to the type of animation needed (coordinates, line offset, etc.)

I have a few concerns, mostly around complexity:

- I feel like it makes the JS interface more confusing. In my opinion, animating something shouldn’t feel like a completely separate action from positioning it - you should just be able to say “and move it smoothly” and have it do that. I worry about needing to nest source components in a specific way (when already the nested source -> layer system can be confusing enough).

- I think adding more native components makes this harder to maintain. For future devs, understanding the native `ShapeSource` implementation as a kind of linear switchboard will be easier.

To minimize complexity in the native implementations, we can simply refactor out utilities, something like `animateTo(value, duration, fps)`. This would dramatically shrink those files.

To simplify the JS interface (i.e. not allow invalid prop combinations), it would be simple to create composed components like `LineStringShapeSource`, which would return `ShapeSource` with only the appropriate values set.

In short, I think we should make animation feel more like an optional setting, instead of a different kind of functionality. Type restrictions would be better handled with composition.

What’s your feeling about all this?

**Comment by mfazekas** [comment]
> @mfazekas Interesting idea. It does have the benefit of keeping the animation component specific to the type of animation needed (coordinates, line offset, etc.)
> 
> I have a few concerns, mostly around complexity:
> 
> * I feel like it makes the JS interface more confusing. In my opinion, animating something shouldn’t feel like a completely separate action from positioning it - you should just be able to say “and move it smoothly” and have it do that. I worry about needing to nest source components in a specific way (when already the nested source -> layer system can be confusing enough).
> * I think adding more native components makes this harder to maintain. For future devs, understanding the native `ShapeSource` implementation as a kind of linear switchboard will be easier.
> 
> To minimize complexity in the native implementations, we can simply refactor out utilities, something like `animateTo(value, duration, fps)`. This would dramatically shrink those files.
> 
> To simplify the JS interface (i.e. not allow invalid prop combinations), it would be simple to create composed components like `LineStringShapeSource`, which would return `ShapeSource` with only the appropriate values set.
> 
> In short, I think we should make animation feel more like an optional setting, instead of a different kind of functionality. Type restrictions would be better handled with composition.
> 
> What’s your feeling about all this?

I do not want to add animation to the interface of ShapeSource. There can be many animations possible, there could be a rotation animation, a translation animation, route animation etc. For me animation is a different kind of functionality, and we should be having it outside ShapeSource both in terms of implementation and the interface.

Also an animation might be source for multiple ShapSource. Like the same route animation can output the route as a LineString to one ShapeSource, but also the end point as a Point to another ShapeSource.

It's a valid point that embedding can be constraining as it might force some layer ordering. Another alternative would be similar to the way react-native and reanimated handle animations.

```jsx
const shape = useMemo(() => new TransitionShapeAnimator(originalShape), []);
useEffect(() => { shape.moveBy(x,y) }, []);
...
<MapView>
   <ShapeSource shape={shape} ... />
</MapView>
```

**Comment by mfazekas** [comment]
> After thinking about it more, I generally agree with you. Of the two options you laid out, I prefer the second one, as I think it's easier to understand and use.
> 
> So, for something specific, like animating the offsets of a line, what do you think about this example usage:
> 
> ```ts
> const animatedShape = useMemo(() => {
>   return new ShapeTransitionLineOffsets({
>     shape: routeLineString,
>     initialStartOffset: 0,
>     initialEndOffset: 0,
>   });
> }, []);
> 
> useEffect(() => {
>   animatedShape.current.setShape(routeLineString);
> }, [routeLineString]);
> 
> useEffect(() => {
>   animatedShape.current.changeStartOffset(startOffset, { duration: 800 });
> }, [startOffset]);
> ```
> 
> The only concern I have is that it requires some boilerplate. A more concise (and React-y) alternative might be to ditch the class instance idea in favor of a pure interface:
> 
> ```ts
> const animatedShape = useMemo((): ShapeTransitionLineOffsets => {
>   return {
>     shape: routeLineString,
>     startOffset,
>     endOffset: 0,
>     duration: 800
>   };
> }, [routeLineString, startOffset]);
> ```

Thanks, I'd prefer the first one. Iideally in longer term we could have method like stopAnimation to stop an animation and return the last state. I don't see that's possible with the second option.

Instead of `new ShapeTransitionLineOffsets` we could do a `createShapeTransitionOffsets`

**Comment by naftalibeder** [comment]
@mfazekas Would you have any objection to invoking functions on the component ref, like

```ts
  const sourceAnimator = useRef<ShapeSourceAnimator>(null);

  useEffect(() => {
    sourceAnimator.current.setLineStartOffset(10, { duration: 2000 });
    // Could also do something like:
    // sourceAnimator.current.animate({ type: 'lineOffsets', startOffset: 10, duration: 2000 });
  }, []);

  return (
    <ShapeSource ref={sourceAnimator}>
      {...}
    </ShapeSource>
  );
```

Pros:
 - It matches the interface we have for animations/transitions on `MapView`, `Camera`, etc.
 - It's a lot simpler than creating an animator class, which requires setting up a global manager instance and dealing with maintaining references to all nodes instances.
 - It still allows for `ref.stopAnimation()`, etc.

Cons:
 - Honestly, not sure I see any ¯\\_(ツ)_/¯

**Comment by mfazekas** [comment]
> @mfazekas Would you have any objection to invoking functions on the component ref, like
> 
> ```ts
>   const sourceAnimator = useRef<ShapeSourceAnimator>(null);
> 
>   useEffect(() => {
>     sourceAnimator.current.setLineStartOffset(10, { duration: 2000 });
>     // Could also do something like:
>     // sourceAnimator.current.animate({ type: 'lineOffsets', startOffset: 10, duration: 2000 });
>   }, []);
> 
>   return (
>     <ShapeSource ref={sourceAnimator}>
>       {...}
>     </ShapeSource>
>   );
> ```
> 
> Pros:
> 
> * It matches the interface we have for animations/transitions on `MapView`, `Camera`, etc.
> * It's a lot simpler than creating an animator class, which requires setting up a global manager instance and dealing with maintaining references to all nodes instances.
> * It still allows for `ref.stopAnimation()`, etc.
> 
> Cons:
> 
> * Honestly, not sure I see any ¯\_(ツ)_/¯

I don't like this a shape animation still would be still strongly coupled with shapeSource both in terms of interface and in implementation. 

Also the example of having an animator animating the route then having 2 shape sources connected to that animator, one getting the route itself (LineString) other just the endpoint (Car symbol for example) would not be possible this way.

I've updated my POC #3012. The interface there is:

```jsx
animatedRoute = new RouteShapeAnimator():
() => {
   animatedRoutes.animate(...);
}

...
<ShapeSource shape={animatedRoute}>
...
</ShapeSource>

```

Yes the it's a bit more complex implementation, but adding different animations are much easier both in terms of implementation and interface.

**Comment by naftalibeder** [comment]
I do like your update to the POC. I guess I don’t mind the added complexity of the native singleton tracking animators, if the added benefit is worth the trade-off - still not sure though. 

Could you post an example of how two ShapeSources could use different parts of the same animator? Would an animator have getter methods like `getLineSlice(start, end)`?

**Comment by mfazekas** [comment]
> I do like your update to the POC. I guess I don’t mind the added complexity of the native singleton tracking animators, if the added benefit is worth the trade-off - still not sure though.
> 
> Could you post an example of how two ShapeSources could use different parts of the same animator’s shape? Would an animator have getter methods like `getLineSlice(start, end)`?

Something like this:
```
let routeAnimator = useMemo(() =>  new RouteAnimator(route), []);
let routeEndAnimator = useMemo(() => new RouteEndAnimator(routeAnimator));


return (
  <ShapeSource shape={routeAnimator}>
     <LineLayer ... />
  </ShapeSource>
  <ShapeSource shape={rounteEndAnimator}>
     <CircleLayer .../>
     <SymbolLayer .../>
  </ShapeSource>
);

```

But instead of `new RouteEndAnimator(routeAnimator)` we could also implement as  `routeAnimator.getEndpointAnimator()`

**Review by Isoftware23 — APPROVED** [review]

**Comment by naftalibeder** [comment]
@mfazekas I was wondering if there was any possibility of integrating something like what's in this PR right now - potentially with `experimental` tags on the animation properties? I haven't had the free time to refactor this into the `ShapeAnimator` architecture, and I don't just want to leave this PR open indefinitely.

if you're not comfortable with that, I can close this PR and pull the functionality out into some custom components that I'll maintain in my own fork. That would be OK with me, although of course it's preferable to be fully synced with upstream. Please let me know what you think.

**Comment by mfazekas** [comment]
@naftalibeder I'm sorry, I do like the feature but I don't want to add it in current form, thanks for the understanding.

If I have time I'll try to work on the shape animators, so adding this is easier

**Comment by ipoogleduck** [comment]
Any update on this? Hoping to add an animated line to my map (as shown in the PR examples) and wondering if there are any new methods or workarounds to make this function.

**Comment by naftalibeder** [comment]
@ipoogleduck This was actually implemented a while back: https://github.com/rnmapbox/maps/pull/3281
